### PR TITLE
First attempt at fixing concurrent file uploads queuing

### PR DESCRIPTION
**Pull request by ianopolous**

### CODE REVIEW

File: src/peergos/server/tests/UserTests.java [hunk 1]
```diff
@@ -583,16 +583,21 @@ public void cleanupFailedUpload() throws Exception {
...
         } catch (Exception e) {}
         int during = context.getTotalSpaceUsed(context.signer.publicKeyHash, context.signer.publicKeyHash).get().intValue();
         Assert.assertTrue("One chunk uploaded", during > 5 * 1024*1024);
 
-        Set<Transaction> pending = context.getTransactionService().getOpenTransactions().join();
-        pending.forEach(t -> context.getTransactionService().clearAndClose(t).join());
+        network.synchronizer.applyComplexUpdate(userRoot.owner(), transactions.getSigner().join(),
+                (s, committer) -> {
+                    Set<Transaction> pending = transactions.getOpenTransactions(s).join();
+                    return Futures.reduceAll(pending, s, (v, t) -> transactions.clearAndClose(v, committer, t), (a, b) -> b);
```
Comment (kevodwyer) [inline src/peergos/server/tests/UserTests.java]: too many single letter vars. perhaps make t -> txn and v -> snapshotVersion

Comment (ianopolous) [reply src/peergos/server/tests/UserTests.java]: fixed

File: src/peergos/shared/user/UserContext.java [hunk 2]
```diff
@@ -1028,8 +1042,14 @@ private static void ensureAllowedToShare(FileWrapper file, String ourname, boole
...
-    private synchronized CompletableFuture<TrieNode> addRootEntryPointAndCommit(TrieNode root, EntryPoint entry) {
-        return writeSynchronizer.applyUpdate(signer.publicKeyHash, signer, (wd, tid) -> {
+    private static CompletableFuture<TrieNode> addRootEntryPointAndCommit(TrieNode root,
+                                                                          EntryPoint entry,
+                                                                          String username,
+                                                                          SigningPrivateKeyAndPublicHash owner,
+                                                                          SymmetricKey rootKey,
+                                                                          NetworkAccess network,
+                                                                          Crypto crypto) {
+        return network.synchronizer.applyUpdate(owner.publicKeyHash, owner, (wd, tid) -> {
```
Comment (kevodwyer) [inline src/peergos/shared/user/UserContext.java]: use writeSynchronizer var

File: src/peergos/shared/user/UserContext.java [hunk 3]
```diff
@@ -559,34 +572,35 @@ public boolean isJavascript() {
...
             // and authorise the writer key
-            SigningPrivateKeyAndPublicHash writerWithHash =
+            SigningPrivateKeyAndPublicHash writerPair =
                     new SigningPrivateKeyAndPublicHash(writerHash, writer.secretSigningKey);
             WritableAbsoluteCapability rootPointer =
                     new WritableAbsoluteCapability(owner.publicKeyHash, writerHash, rootMapKey, rootRKey, rootWKey);
-            EntryPoint entry = new EntryPoint(rootPointer, this.username);
-            return addOwnedKeyAndCommit(writerWithHash)
+            EntryPoint entry = new EntryPoint(rootPointer, directoryName);
+            return network.synchronizer.applyUpdate(owner.publicKeyHash, owner,
```
Comment (kevodwyer) [inline src/peergos/shared/user/UserContext.java]: use writeSynchronizer var

File: src/peergos/shared/user/UserContext.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/shared/user/UserContext.java]: it's a static method now.

File: src/peergos/shared/user/UserContext.java (hunk 3 above)
Comment (ianopolous) [reply src/peergos/shared/user/UserContext.java]: it's a static method now

**Pull request merged**